File: logic/python/stack.py
"""
Quản lý lịch sử mật khẩu cũ của tài khoản theo CTDL Stack.

Version: Python
Last edit: 2026-04-25
Author: 25520260
"""

from datetime import datetime


class PasswordStack:
    """
    Class dùng để quản lý lịch sử mật khẩu cũ của 1 tài khảon.

    Mỗi phần tử trong stack gồm:
    {
        "password": mật khẩu cũ,
        "changedAt": thời gian đổi mật khẩu
    }
    """

    def __init__(self, max_size=20):
        self.items = []
        self.max_size = max_size

    def size(self):
        """Trả về số lượng phần tử hiện có.
        Args:
            self
        Returns:
            len(self.items)
        """
        return len(self.items)

    def is_empty(self):
        """Kiểm tra stack có rỗng không.
        Args:
            self
        Returns:
            bool
        """
        return len(self.items) == 0

    def push(self, password):
        """
        Thêm mật khẩu cũ vào stack.
        Khi đổi mật khẩu mới thì mật khẩu cũ sẽ được lưu vào đây.
        Args:
            self
            password
        Returns:
            None
        Note:
            self.items.insert(0, new_item) --- thêm phần tử mới vào đầu stack
            self.items.pop() --- xóa phần tử cũ nhất nếu vượt quá giới hạn
        """
        if password == "":
            return

        time_now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

        new_item = {
            "password": password,
            "changedAt": time_now
        }

        self.items.insert(0, new_item)

        if len(self.items) > self.max_size:
            self.items.pop()

    def pop(self):
        """
        Lấy và xóa mật khẩu gần nhất.
        Args:
            self
        Returns:
            password gần nhất hoặc None nếu stack rỗng
        """
        if self.is_empty():
            return None

        return self.items.pop(0)

    def peek(self):
        """
        Xem mật khẩu gần nhất nhưng không xóa.
        Args:
            self
        Returns:
            password gần nhất hoặc None nếu stack rỗng
        """
        if self.is_empty():
            return None

        return self.items[0]

    def is_duplicate(self, password):
        """
        Kiểm tra mật khẩu đã từng dùng chưa.
        Args:
            self
            password
        Returns:
            bool
        """
        for item in self.items:
            if item["password"] == password:
                return True

        return False

    def restore_at(self, index, current_password):
        """
        Khôi phục mật khẩu cũ tại vị trí index.
        Args:
            self
            index
            current_password
        Returns:
            password đã khôi phục hoặc None nếu vị trí không hợp lệ
        """
        if index < 0 or index >= len(self.items):
            return None

        old_item = self.items.pop(index)

        self.push(current_password)

        return old_item["password"]

    def clear(self):
        self.items = []

    def to_array(self):
        """Trả về danh sách lịch sử.
        Args:
            self
        Returns:
            list
        """
        return self.items


def main():
    history = PasswordStack(5)

    current_password = "abc123"

    print("Mật khẩu hiện tại:", current_password)
    print("Lịch sử ban đầu:", history.to_array())

    history.push(current_password)
    current_password = "abc456"

    print("\nSau khi đổi mật khẩu:")
    print("Mật khẩu hiện tại:", current_password)
    print("Lịch sử:", history.to_array())

    check = history.is_duplicate("abc123")
    print("\nabc123 đã từng dùng chưa?", check)

    print("Mật khẩu cũ gần nhất:", history.peek())

    restore = history.restore_at(0, current_password)
    current_password = restore

    print("\nSau khi khôi phục:")
    print("Mật khẩu hiện tại:", current_password)
    print("Lịch sử:", history.to_array())


if __name__ == "__main__":
    main()

File: logic/stack.js
/**
 * @file stack.js
 * @description Quản lý lịch sử mật khẩu cũ của tài khoản theo CTDL Stack.
 *
 * Mỗi phần tử trong stack có dạng:
 * {
 *   password: "mật-khẩu-cũ",
 *   changedAt: "thời-gian-đổi-mật-khẩu"
 * }
 */

/**
 * Lớp quản lý lịch sử mật khẩu cũ của một tài khoản.
 */
export class PasswordStack {
  /**
   * @param {Array<{password: string, changedAt: string}>} [initialItems=[]]
   * @param {number} [maxSize=20]
   */
  constructor(initialItems = [], maxSize = 20) {
    this.items = Array.isArray(initialItems) ? initialItems.slice(0, maxSize).map((item) => ({ ...item })) : [];
    this.maxSize = maxSize;
  }

  /**
   * Trả về số lượng phần tử hiện có.
   *
   * @returns {number}
   */
  get size() {
    return this.items.length;
  }

  /**
   * Kiểm tra stack có rỗng không.
   *
   * @returns {boolean}
   */
  isEmpty() {
    return this.items.length === 0;
  }

  /**
   * Thêm mật khẩu cũ vào đầu stack.
   *
   * @param {string} password
   * @param {string} [changedAt]
   * @returns {PasswordStack}
   */
  push(password, changedAt = new Date().toISOString()) {
    if (!password) {
      return this;
    }

    const newItem = {
      password,
      changedAt
    };

    this.items.unshift(newItem);

    if (this.items.length > this.maxSize) {
      this.items.pop();
    }

    return this;
  }

  /**
   * Lấy và xóa mật khẩu cũ gần nhất.
   *
   * @returns {{password: string, changedAt: string} | undefined}
   */
  pop() {
    if (this.isEmpty()) {
      return undefined;
    }

    return this.items.shift();
  }

  /**
   * Xem mật khẩu cũ gần nhất nhưng không xóa.
   *
   * @returns {{password: string, changedAt: string} | undefined}
   */
  peek() {
    if (this.isEmpty()) {
      return undefined;
    }

    return this.items[0];
  }

  /**
   * Kiểm tra mật khẩu đã từng xuất hiện trong lịch sử hay chưa.
   *
   * @param {string} password
   * @returns {boolean}
   */
  isDuplicate(password) {
    for (const item of this.items) {
      if (item.password === password) {
        return true;
      }
    }

    return false;
  }

  /**
   * Khôi phục mật khẩu cũ tại vị trí index.
   *
   * @param {number} index
   * @param {string} currentPassword
   * @returns {{restoredPassword: string, changedAt: string} | null}
   */
  restoreAt(index, currentPassword) {
    if (index < 0 || index >= this.items.length) {
      return null;
    }

    const oldItem = this.items.splice(index, 1)[0];
    this.push(currentPassword);

    return {
      restoredPassword: oldItem.password,
      changedAt: oldItem.changedAt
    };
  }

  /**
   * Xóa toàn bộ lịch sử.
   *
   * @returns {PasswordStack}
   */
  clear() {
    this.items = [];
    return this;
  }

  /**
   * Trả về bản sao của lịch sử.
   *
   * @returns {Array<{password: string, changedAt: string}>}
   */
  toArray() {
    return this.items.map((item) => ({ ...item }));
  }

  /**
   * Tạo PasswordStack từ mảng dữ liệu có sẵn.
   *
   * @param {Array<{password: string, changedAt: string}>} data
   * @param {number} [maxSize=20]
   * @returns {PasswordStack}
   */
  static fromArray(data, maxSize = 20) {
    return new PasswordStack(data, maxSize);
  }
}
